### PR TITLE
Introduced `EventsDriven`

### PR DESCRIPTION
**Pull request by KickItLikeShika**
As explained here: https://github.com/pytorch/ignite/wiki/Roadmap#refactor-engine
This PR will separate the events registration code from the other code.

Check list:

- [x] New tests are added (if a new feature is added)
- [x] New doc strings: description and/or example code are in RST format
- [x] Documentation is updated (if required)

Code adapted from https://github.com/vfdev-5/ignite/tree/eventdriven

### CODE REVIEW

**Comment by KickItLikeShika** [comment]
Please review this as soon as you have time, @sdesrozis @trsvchn @vfdev-5. Thanks.

File: ignite/base/mixins.py [hunk 1]
```diff
@@ -22,3 +28,262 @@ def load_state_dict(self, state_dict: Mapping) -> None:
         opts = [k in state_dict for k in self._state_dict_one_of_opt_keys]
         if len(opts) > 0 and ((not any(opts)) or (all(opts))):
             raise ValueError(f"state_dict should contain only one of '{self._state_dict_one_of_opt_keys}' keys")
+
+
+class EventsDriven:
```
Comment (sdesrozis) [inline ignite/base/mixins.py:33]: Even if it’s a mixin, it’s an important library’s feature, as Metric for instance. It would deserve a specific file. What do you think?

File: ignite/base/mixins.py [hunk 2]
```diff
@@ -22,3 +28,262 @@ def load_state_dict(self, state_dict: Mapping) -> None:
...
+    """
+
+    def __init__(self) -> None:
+        self._event_handlers = defaultdict(list)  # type: Dict[Any, List]
+
+        self._allowed_events = []  # type: List[EventEnum]
+        self._allowed_events_counts = {}  # type: Dict[Union[str, EventEnum], int]
+
+        self.last_event_name = None  # type: Optional[Events]
+        self.logger = logging.getLogger(__name__ + "." + self.__class__.__name__)
```
Comment (sdesrozis) [inline ignite/base/mixins.py:45]: Could we use an ignite logger ?

File: ignite/base/mixins.py [hunk 3]
```diff
@@ -22,3 +28,262 @@ def load_state_dict(self, state_dict: Mapping) -> None:
...
+
+        return decorator
+
+    def _fire_event(self, event_name: Any, *event_args: Any, **event_kwargs: Any) -> None:
+        """Execute all the handlers associated with given event.
+
+        This method executes all handlers associated with the event
+        `event_name`. Optional positional and keyword arguments can be used to
+        pass arguments to **all** handlers added with this event. These
+        arguments updates arguments passed using :meth:`~ignite.base.EventsDriven.add_event_handler`.
```
Comment (sdesrozis) [inline ignite/base/mixins.py:202]:
```suggestion
        arguments updates arguments passed using :meth:`~ignite.base.mixins.EventsDriven.add_event_handler`.
```
As L185 ?

File: ignite/engine/engine.py [hunk 4]
```diff
@@ -290,39 +268,7 @@ def execute_something():
...
-
-    @staticmethod
-    def _assert_non_filtered_event(event_name: Any) -> None:
-        if (
-            isinstance(event_name, CallableEventWithFilter)
-            and event_name.filter != CallableEventWithFilter.default_event_filter
-        ):
-            raise TypeError(
-                "Argument event_name should not be a filtered event, " "please use event without any event filtering"
-            )
```
Comment (sdesrozis) [inline ignite/engine/engine.py:325]: Useless ?

File: ignite/engine/engine.py [hunk 5]
```diff
@@ -332,23 +278,7 @@ def has_event_handler(self, handler: Callable, event_name: Optional[Any] = None)
...
-                if self._compare_handlers(handler, h):
-                    return True
-        return False
-
-    @staticmethod
-    def _compare_handlers(user_handler: Callable, registered_handler: Callable) -> bool:
-        if hasattr(registered_handler, "_parent"):
-            registered_handler = registered_handler._parent()  # type: ignore[attr-defined]
-        return registered_handler == user_handler
+        return super(Engine, self).has_event_handler(handler, event_name=event_name)
```
Comment (sdesrozis) [inline ignite/engine/engine.py:281]: Same here, useless ?

File: ignite/engine/engine.py [hunk 6]
```diff
@@ -394,35 +314,7 @@ def execute_something():
...
-            **event_kwargs: optional keyword args to be passed to all handlers.
-
-        """
-        self.logger.debug(f"firing handlers for event {event_name}")
-        self.last_event_name = event_name
-        for func, args, kwargs in self._event_handlers[event_name]:
-            kwargs.update(event_kwargs)
-            first, others = ((args[0],), args[1:]) if (args and args[0] == self) else ((), args)
-            func(*first, *(event_args + others), **kwargs)
+        return super(Engine, self).on(event_name, *args, **kwargs)
```
Comment (sdesrozis) [inline ignite/engine/engine.py:317]: Same here.

File: ignite/engine/engine.py [hunk 7]
```diff
@@ -445,8 +337,7 @@ def fire_event(self, event_name: Any) -> None:
                 :meth:`~ignite.engine.engine.Engine.register_events`.
 
         """
-        self._assert_allowed_event(event_name)
-        return self._fire_event(event_name)
+        super(Engine, self).fire_event(event_name)
```
Comment (sdesrozis) [inline ignite/engine/engine.py:340]: And finally here.

**Comment by sdesrozis** [comment]
@KickItLikeShika Thanks for this PR! It's a tricky part, and good for you for being comfortable enough with it!

I left a few comments and some tests must be fixed.

File: ignite/engine/engine.py (hunk 5 above)
Comment (KickItLikeShika) [reply ignite/engine/engine.py:281]: Yes, all the functions left in the Engine is left for the keeping the docs.

File: ignite/base/mixins.py (hunk 3 above)
Comment (KickItLikeShika) [reply ignite/base/mixins.py:202]: Yes that's right, thanks!

File: ignite/base/mixins.py (hunk 2 above)
Comment (KickItLikeShika) [reply ignite/base/mixins.py:45]: It's the same logger used in the Engine now, do we want to change it? https://github.com/pytorch/ignite/blob/master/ignite/engine/engine.py#L126

File: ignite/base/mixins.py (hunk 1 above)
Comment (KickItLikeShika) [reply ignite/base/mixins.py:33]: I agree with that, thanks!

**Comment by KickItLikeShika** [comment]
@sdesrozis can you please point out to the tests that should be fixed? only the there is one failing test that should be fixed, usually it was working fine without any updates, see here #2106, so I'm a bit surprised with this failure, do you thing there is other tests that should be updated?

**Comment by vfdev-5** [comment]
@KickItLikeShika there are many failing tests (13) which should be fixed before merging.

**Comment by KickItLikeShika** [comment]
@vfdev-5 Yes I mean just 2 unit tests should be fixed

File: ignite/base/mixins.py (hunk 2 above)
Comment (sdesrozis) [reply ignite/base/mixins.py]: That's the point I would like to figure out.

File: .github/workflows/tpu-tests.yml [hunk 8]
```diff
@@ -64,14 +64,6 @@ jobs:
           pip install -r requirements-dev.txt
           python setup.py install
 
-      # https://github.com/pytorch/ignite/issues/1884
```
Comment (vfdev-5) [inline .github/workflows/tpu-tests.yml:67]: Please revert this change. I'll update `refactor-engine` branch separately.

File: ignite/engine/engine.py [hunk 9]
```diff
@@ -290,39 +269,7 @@ def execute_something():
...
-    @staticmethod
-    def _assert_non_filtered_event(event_name: Any) -> None:
-        if (
-            isinstance(event_name, CallableEventWithFilter)
-            and event_name.filter != CallableEventWithFilter.default_event_filter
-        ):
-            raise TypeError(
-                "Argument event_name should not be a filtered event, " "please use event without any event filtering"
-            )
+        return super(Engine, self).add_event_handler(event_name, handler, *args, **kwargs)
```
Comment (sdesrozis) [inline ignite/engine/engine.py:272]: @KickItLikeShika you resolved the comment, yet you didn’t answer.

**Comment by sdesrozis** [comment]
@KickItLikeShika error in the docstring of EventDriver.

File: ignite/base/events_driven.py [hunk 10]
```diff
@@ -0,0 +1,267 @@
...
+from typing import Any, Callable, Dict, List, Mapping, Optional, Union
+
+from ignite.engine.events import CallableEventWithFilter, EventEnum, Events, EventsList, RemovableEventHandle
+from ignite.engine.utils import _check_signature
+
+
+class EventsDriven:
+    """Base class for events-driven engines without state.
+
+    """
```
Comment (vfdev-5) [inline ignite/base/events_driven.py:14]: If we make this class public and expose it in the docs. We need to make this docstring better and provide examples

**Comment by vfdev-5** [comment]
@KickItLikeShika I thought we agreed on the first PR to do only:

> - introducing EventsDriven and other mixins classes without changing anything in Engine

Here you are already modifying `Engine`.

File: tests/ignite/base/test_events_driven.py [hunk 11]
```diff
@@ -0,0 +1,238 @@
...
+
+class ABCEvents(EventEnum):
+    A_EVENT = "a_event"
+    B_EVENT = "b_event"
+    C_EVENT = "c_event"
+
+
+# ! There are some updates to be done in EventsDriven
+# ! uncomment this when EventsDriven is updated in the upcoming PR
+# def test_events_driven_basics():
```
Comment (vfdev-5) [inline tests/ignite/base/test_events_driven.py:29]: Which updates you are talking about here? I think this test file should contain all tests. If it is not the case, maybe something is wrong

File: ignite/engine/engine.py (hunk 9 above)
Comment (KickItLikeShika) [reply ignite/engine/engine.py:272]: Very sorry about that, but I answered your question here https://github.com/pytorch/ignite/pull/2107#discussion_r669973702, I thought you saw it and agreed so I resolved the comment, again sorry.
I think keeping the full docs in the Engine docs and minimized docs in the base class if something nice, what do you think?

File: ignite/base/events_driven.py (hunk 10 above)
Comment (KickItLikeShika) [reply ignite/base/events_driven.py:14]: I agree with that, thanks!

File: tests/ignite/base/test_events_driven.py (hunk 11 above)
Comment (KickItLikeShika) [reply tests/ignite/base/test_events_driven.py:29]: Sorry I should have made that more clear in the comments, currently we are fetching the event by the same old way in the (using state) `_handler_wrapper` https://github.com/pytorch/ignite/blob/44a64d6d8def637ae636471faf1487a18d209456/ignite/base/events_driven.py#L49, which is will make this test fail. Why don't we update it and fetch the event using `_allowed_event_counts` as we should do in the upcoming updates? using `_allowed_event_counts` without the other necessary updates along with it will make `test_deterministic` fails, so I commented it for now, and as soon as we update the engine and `_handler_wrapper` I will uncomment this.

**Comment by KickItLikeShika** [comment]
@vfdev-5 can you please point out to which modification in the engine you mean? I haven't changed anything in the engine except the necessary for introducing EventsDriven

File: tests/ignite/base/test_events_driven.py (hunk 11 above)
Comment (vfdev-5) [reply tests/ignite/base/test_events_driven.py:29]: > Why don't we update it and fetch the event using _allowed_event_counts as we should do in the upcoming updates?

Yes, this is also my question :) Ahmed, it is your PR


I commented before to remove EventsDriven from Engine and in this PR we should just introduce EventsDriven...
All tests should be present and passing.

File: ignite/engine/engine.py [hunk 12]
```diff
@@ -1,23 +1,23 @@
...
+from ignite.base.mixins import Serializable
+from ignite.engine.events import EventEnum, Events, RemovableEventHandle
+from ignite.engine.state import State
 from ignite.engine.utils import _check_signature, _to_hours_mins_secs
 
 __all__ = ["Engine"]
 
 
-class Engine(Serializable):
+class Engine(Serializable, EventsDriven):
```
Comment (vfdev-5) [inline ignite/engine/engine.py:20]: @KickItLikeShika Engine should not inherit from `EventsDriven` in this PR

File: tests/ignite/base/test_events_driven.py (hunk 11 above)
Comment (KickItLikeShika) [reply tests/ignite/base/test_events_driven.py:29]: > using _allowed_event_counts without the other necessary updates along with it will make test_deterministic fails, so I commented it for now, and as soon as we update the engine and _handler_wrapper I will uncomment this.

That's the answer for the question. I understand your point now, I will revert the engine updates.

File: ignite/engine/events.py [hunk 13]
```diff
@@ -348,81 +347,6 @@ def __or__(self, other: Union[Events, CallableEventWithFilter]) -> "EventsList":
         return self
 
 
-class State:
```
Comment (vfdev-5) [inline ignite/engine/events.py:351]: @KickItLikeShika let's revert changes with the State as well. Thanks

Comment (KickItLikeShika) [reply ignite/engine/events.py:351]: What about keeping it like this for now? My point is in this PR we don't do huge changes that requires a lot of reviews from your side, so I think moving State in this PR will prevent you from more headaches in next PR, and the next PR will be more focused on the important changes, what do you think?

Comment (vfdev-5) [reply ignite/engine/events.py:351]: Let's do things by logical steps

Comment (KickItLikeShika) [reply ignite/engine/events.py:351]: I see your point, I will revert state updates.

File: ignite/base/events_driven.py [hunk 14]
```diff
@@ -0,0 +1,319 @@
...
+    ) -> None:
+        """Add events that can be fired.
+
+        Args:
+            event_names: Defines the name of the event being supported. New events can be a str
+                or an object derived from :class:`~ignite.engine.events.EventEnum`. See example below.
+            event_to_attr: A dictionary to map an event to a state attribute.
+        """
+        if not (event_to_attr is None or isinstance(event_to_attr, dict)):
+            raise ValueError(f"Expected event_to_attr to be dictionary. Got {type(event_to_attr)}.")
```
Comment (vfdev-5) [inline ignite/base/events_driven.py:74]: @KickItLikeShika please add tests to cover uncovered cases.

File: tests/ignite/base/test_mixins.py [hunk 15]
```diff
@@ -1,6 +1,6 @@
 import pytest
 
-from ignite.base import Serializable
+from ignite.base.mixins import Serializable
```
Comment (vfdev-5) [inline tests/ignite/base/test_mixins.py:3]: I think this can be reverted.

**Review by vfdev-5 — COMMENTED** [review]
@KickItLikeShika Thanks for the updates. Remains few things and we can merge it

**Review by vfdev-5 — APPROVED** [review]
Thanks @KickItLikeShika !

File: ignite/base/events_driven.py [hunk 16]
```diff
@@ -0,0 +1,319 @@
...
+from typing import Any, Callable, Dict, List, Mapping, Optional, Union
+
+from ignite.engine.events import CallableEventWithFilter, EventEnum, Events, EventsList, RemovableEventHandle
+from ignite.engine.utils import _check_signature
+
+
+class EventsDriven:
+    """Base class for events-driven engines without state. This class is mainly
+    respoinsible for registring events and fire events, also keep track of
+    the allowed events, and how many time they have been fired.
```
Comment (vfdev-5) [inline ignite/base/events_driven.py:14]:
```suggestion
    responsible for registering events and triggering events, it also keeps track of
    the allowed events, and how many times they have been triggered.
```
typos

File: ignite/base/events_driven.py [hunk 17]
```diff
@@ -0,0 +1,319 @@
...
+
+    Attributes:
+        last_event_name: last event name triggered by the engine.
```
Comment (vfdev-5) [inline ignite/base/events_driven.py:17]:
```suggestion
        last_event_name: last event name triggered.
```

File: ignite/base/events_driven.py [hunk 18]
```diff
@@ -0,0 +1,319 @@
...
+
+    Examples:
```
Comment (vfdev-5) [inline ignite/base/events_driven.py:19]: This does not look good:
- https://deploy-preview-2107--pytorch-ignite-preview.netlify.app/generated/ignite.base.events_driven.eventsdriven#ignite.base.events_driven.EventsDriven

Comment (KickItLikeShika) [reply ignite/base/events_driven.py:19]: I agree that doesn't look very neat, but that's the same way we render examples in the entire codebase, should I find a way to fix it here? or leave it?

Comment (vfdev-5) [reply ignite/base/events_driven.py:19]: We may do that later everywhere. Could you please open an issue about that. Thanks !
